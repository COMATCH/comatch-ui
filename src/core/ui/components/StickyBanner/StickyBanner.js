import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './StickyBanner.scss';

/**
 * A banner that appear on the top of the browser's window the follows the user as they scroll down.
 */

export class StickyBanner extends Component {
    static propTypes = {
        /**
         * Value in pixels. If defined determines at which distance
         * from the document's top the `StickyBanner` will become visible.
         */
        scrollTopVisibilityThreshold: PropTypes.number,
        children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
        /**
         * Additional class names
         */
        className: PropTypes.string,
    };

    static defaultProps = {
        scrollTopVisibilityThreshold: null,
        children: null,
        className: null,
    };

    state = {
        isShown: false,
    };

    componentDidMount() {
        const { scrollTopVisibilityThreshold } = this.props;

        if (scrollTopVisibilityThreshold) {
            // If scrollTopVisibilityThreshold is defined, then we must attach an event
            // handler on document to monitor when to hide/show the sticky bar.
            const htmlEl = document.querySelector('html');
            this.documentScrollListener = () => {
                const { isShown } = this.state;
                if (htmlEl.scrollTop >= scrollTopVisibilityThreshold && !isShown) {
                    this.setState({ isShown: true });
                } else if (htmlEl.scrollTop < scrollTopVisibilityThreshold && isShown) {
                    this.setState({ isShown: false });
                }
            };
            document.addEventListener('scroll', this.documentScrollListener);
        }
    }

    componentWillUnmount() {
        const { scrollTopVisibilityThreshold } = this.props;

        if (scrollTopVisibilityThreshold && this.documentScrollListener) {
            // remove event listener and clean up
            document.removeEventListener('scroll', this.documentScrollListener);
            this.documentScrollListener = null;
        }
    }

    render() {
        const { scrollTopVisibilityThreshold, children, className } = this.props;

        const classes = classNames('StickyBanner', className, {
            animated: scrollTopVisibilityThreshold,
            shown: scrollTopVisibilityThreshold && this.state.isShown,
        });
        return (
            <section className={classes}>
                <div className="StickyBanner__content">{children}</div>
            </section>
        );
    }
}
