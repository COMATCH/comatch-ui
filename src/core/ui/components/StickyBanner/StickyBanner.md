
## StickyBanner

From [`StickyBanner.js`](StickyBanner.js)

A banner that appear on the top of the browser's window the follows the user as they scroll down.

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**children** | `Union<Array[]<ReactNode> \| ReactNode>` | `null` | :x: | 
**className** | `String` | `null` | :x: | Additional class names
**scrollTopVisibilityThreshold** | `Number` | `null` | :x: | Value in pixels. If defined determines at which distance from the document's top the `StickyBanner` will become visible.



