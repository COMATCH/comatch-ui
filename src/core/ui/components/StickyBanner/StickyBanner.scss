@import '../../../styles/variables.scss';
@import '../../../styles/mixins.scss';

.StickyBanner {
    // color & border
    background-color: $co-color-white;
    border-bottom: 1px solid $co-color-mid-gray;
    // display
    display: flex;
    justify-content: center;
    // sizing
    width: 100%;
    padding: 10px 0;
    // positioning
    position: fixed;
    left: 0;
    top: 0;
    z-index: 100;

    &__content {
        display: flex;
        justify-content: center;
        max-width: 1184px; // restrict to grid width
    }

    // states
    &.animated {
        transition: transform 250ms ease-out;
        transform: translateY(-100%);

        &.shown {
            transform: translateY(0);
        }
    }
}