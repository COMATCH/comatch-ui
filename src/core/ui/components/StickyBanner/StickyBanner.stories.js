/* eslint-disable import/no-unresolved */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { StickyBanner } from './StickyBanner';

/* eslint-disable react/prop-types */
const Wrapper = ({ scrollTopVisibilityThreshold }) => (
    <div style={{ height: '1500px' }}>
        <StickyBanner scrollTopVisibilityThreshold={scrollTopVisibilityThreshold}>
            hello
        </StickyBanner>
    </div>
);

storiesOf('StickyBanner', module)
    .add('with basic style', () => <Wrapper />)
    .add('becoming visible after scrollTop of HTML from top is 250px', () => (
        <Wrapper scrollTopVisibilityThreshold={250} />
    ));
