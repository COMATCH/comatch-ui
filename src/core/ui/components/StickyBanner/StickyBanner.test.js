import React from 'react';
import renderer from 'react-test-renderer';
import { lorem } from 'faker';
import { mount } from 'enzyme';
import { StickyBanner } from './StickyBanner';

import '../../../config/tests/setup';

describe('StickyBanner', () => {
    it('should render correctly', () => {
        const rendered = renderer.create(<StickyBanner />);
        expect(rendered.toJSON()).toMatchSnapshot();
    });

    it('should render the child', () => {
        const className = lorem.word();
        const rendered = renderer.create(
            <StickyBanner>
                <div className={className}>
                </div>
            </StickyBanner>
        );
        const instance = rendered.root;
        const childNode = instance.findByProps({className});
        expect(childNode.props.className).toEqual(className);
    });

    it('should has animated class if scrollTopVisibilityThreshold passed with value', () => {
        const visibilityThreshold = 250;
        const rendered = renderer.create(
            <StickyBanner
                scrollTopVisibilityThreshold={visibilityThreshold}>
            </StickyBanner>
        );

        const instance = rendered.root;
        expect(instance.children[0].props.className).toEqual('StickyBanner animated');
    });

    it('should display custom class names if passed', () => {
        const className = lorem.word();
        const rendered = renderer.create(
            <StickyBanner
                className={className}>
            </StickyBanner>
        );

        const instance = rendered.root;
        expect(instance.children[0].props.className).toEqual(`StickyBanner ${className}`);
    });

    it('should be shown after the scroll pass the threshold', () => {
        const visibilityThreshold = 250;

        // mock addEventListener on document object
        const mapAdd = {};
        document.addEventListener = jest.fn((event, cb) => {
            mapAdd[event] = cb;
        });

        const wrapper = mount(<StickyBanner scrollTopVisibilityThreshold={visibilityThreshold} />);

        // simulate with value before the passed threshold
        const htmlEl = document.querySelector('html');
        htmlEl.scrollTop = (visibilityThreshold - 1);
        // fire scroll event
        mapAdd.scroll();

        wrapper.update();
        let component = wrapper.find('.StickyBanner');
        expect(document.addEventListener.mock.calls.length).toEqual(1);
        expect(component.prop('className')).toEqual('StickyBanner animated');

        // simulate with value on/after the passed threshold
        htmlEl.scrollTop = visibilityThreshold;
        // fire scroll event
        mapAdd.scroll();

        wrapper.update();
        component = wrapper.find('.StickyBanner');
        expect(component.prop('className')).toEqual('StickyBanner animated shown');

        wrapper.unmount();
    });

    it('should unregister the scroll event on componentWillUnmount', () => {
        // mock removeEventListener on document object
        const mapRemove = {};
        document.removeEventListener = jest.fn((event, cb) => {
            mapRemove[event] = cb;
        });

        // any non-zero value for scrollTopVisibilityThreshold is sufficient to attach the scroll event listener
        const wrapper = mount(<StickyBanner scrollTopVisibilityThreshold={123} />);
        wrapper.unmount();
        expect(document.removeEventListener.mock.calls.length).toEqual(1);
    });
});
