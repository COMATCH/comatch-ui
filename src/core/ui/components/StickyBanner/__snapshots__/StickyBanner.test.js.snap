// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StickyBanner should render correctly 1`] = `
<section
  className="StickyBanner"
>
  <div
    className="StickyBanner__content"
  />
</section>
`;
